src/Storages/MergeTree/MergeTreeCurrentlyRestoringFromBackup.cpp

```diff
@@ -0,0 +1,347 @@
+    {
+        Int64 base_increment_value = getCurrentIncrementValue();
+        std::optional<AllocatedBlockNumbers> allocated_block_numbers;
+
+        /// Check the table has no existing parts if it's necessary.
+        /// If a backup contains mutations the table must have no existing parts,
+        /// otherwise mutations from the backup could be applied to existing parts which is wrong.
+        if (no_parts_check_reason_ != NoPartsCheckReason::NONE)
+            checkNoPartsExist(base_increment_value, allocated_block_numbers, no_parts_check_reason_);
+
+        LOG_INFO(log, "Increasing the increment to allocate block numbers for {} parts and {} mutations",
+                 part_infos_.size(), mutation_infos_.size());
+
+        /// This `do_allocate*` function will be called from calculateBlockNumbersForRestoringMergeTree() below.
+        auto do_allocate_block_numbers = [&](size_t count)
+        {
+            auto lock = storage.lockParts();
+            Int64 first = storage.increment.getMany(count);
+            allocated_block_numbers.emplace(AllocatedBlockNumbers{.first = first, .count = count});
+            return first;
+        };
+
+        /// Recalculate block numbers in `part_infos` and `mutation_infos`.
+        calculateBlockNumbersForRestoringMergeTree(part_infos_, part_names_in_backup_, mutation_infos_, mutation_names_in_backup_, do_allocate_block_numbers);
+
+        /// Check the table has no existing parts again to be sure no parts were added while we were allocating block numbers.
+        if (no_parts_check_reason_ != NoPartsCheckReason::NONE)
+            checkNoPartsExist(base_increment_value, allocated_block_numbers, no_parts_check_reason_);
+
+        LOG_INFO(log, "The increment was incremented by {} to allocate block numbers",
+                 allocated_block_numbers ? allocated_block_numbers->count : 0);
+    }
+
+private:
+    StorageMergeTree & storage;
+    Poco::Logger * log;
+    const CurrentlyRestoringInfo & currently_restoring_info;
+
+    String getTableName() const { return storage.getStorageID().getFullTableName(); }
+
+    struct AllocatedBlockNumbers
+    {
+        Int64 first = 0;
+        size_t count = 0;
+    };
+
+    /// Checks that there no parts exist in the table.
+    /// The function also checks that no parts will be inserted / attached soon due to a concurrent process.
+    void checkNoPartsExist(Int64 base_increment_value, const std::optional<AllocatedBlockNumbers> & ignore_block_numbers, NoPartsCheckReason reason) const
+    {
+        if (reason == NoPartsCheckReason::NONE)
+            return;
+
+        LOG_INFO(log, "Checking that no parts exist in the table before restoring its data (reason: {})", magic_enum::enum_name(reason));
+
+        /// The order of checks is important:
+        /// an INSERT command first allocates a block number, then it attaches it to the table.
+        /// That's why it's more reliable to check first block numbers, then data parts.
```

Comment:
> A concurrent insert can allocate a block number just after we got nullopt from findAnyAllocatedBlockNumber(...), and then it can attach a part just after we got nothing from findAnyPartInStorage().

The function `checkNoPartsExist()` is called two times - see [(1)](https://github.com/ClickHouse/ClickHouse/pull/50230/files#diff-67ce4faec8e617b8ba0699512283ecf11b557ecb259357eabd79bfbb5bed5023R41) and [(2)](https://github.com/ClickHouse/ClickHouse/pull/50230/files#diff-67ce4faec8e617b8ba0699512283ecf11b557ecb259357eabd79bfbb5bed5023R60), so this case will be handled.

Reply:
Yes, checking this second time avoids race conditions, but I mean that the comment is confusing (and probably not entirely correct)